java-checks/src/main/java/org/sonar/java/checks/regex/ImpossibleBoundariesCheck.java

```diff
@@ -69,26 +81,74 @@ public void visitLookAround(LookAroundTree tree) {
       }
     }
 
+    @Override
+    public void visitDisjunction(DisjunctionTree tree) {
+      BoundaryInDisjunctionFounder boundaryInDisjunctionFounder = new BoundaryInDisjunctionFounder();
+      boundaryInDisjunctionFounder.visit(tree);
+      excluded.addAll(boundaryInDisjunctionFounder.foundBoundaries());
+      super.visitDisjunction(tree);
+    }
+
     @Override
     public void visitBoundary(BoundaryTree boundaryTree) {
       switch (boundaryTree.type()) {
         case LINE_START:
         case INPUT_START:
           if (!RegexTreeHelper.canReachWithoutConsumingInput(start, boundaryTree)) {
             reportIssue(boundaryTree, String.format(MESSAGE, "after"), null, Collections.emptyList());
+          } else if (!excluded.contains(boundaryTree) && probablyShouldConsumeInput(start, boundaryTree)) {
+            reportIssue(boundaryTree, String.format(SOFT_MESSAGE, "after"), null, Collections.emptyList());
           }
           break;
         case LINE_END:
         case INPUT_END:
         case INPUT_END_FINAL_TERMINATOR:
           if (!RegexTreeHelper.canReachWithoutConsumingInput(boundaryTree, end)) {
             reportIssue(boundaryTree, String.format(MESSAGE, "before"), null, Collections.emptyList());
+          } else if (!excluded.contains(boundaryTree) && probablyShouldConsumeInput(boundaryTree, end)) {
+            reportIssue(boundaryTree, String.format(SOFT_MESSAGE, "before"), null, Collections.emptyList());
           }
           break;
         default:
           // Do nothing
       }
     }
+
+    private boolean probablyShouldConsumeInput(AutomatonState start, AutomatonState goal) {
+      return  canReachWithConsumingInput(start, goal, new HashSet<>());
+    }
   }
 
+  private static class BoundaryInDisjunctionFounder extends RegexBaseVisitor {
+    private final Set<BoundaryTree> foundBoundaries = new HashSet<>();
+
+    @Override
+    public void visitBoundary(BoundaryTree boundaryTree) {
+      foundBoundaries.add(boundaryTree);
+    }
+
+    public Set<BoundaryTree> foundBoundaries() {
+      return new HashSet<>(foundBoundaries);
+    }
+  }
+
+  private boolean canReachWithConsumingInput(AutomatonState start, AutomatonState goal, Set<AutomatonState> visited) {
+    if (start == goal 
+      || start instanceof LookAroundTree 
```

Comment:
Returning false when encountering a LookAroundTree is wrong. Given the regex `a(?=.)bc`, the path from a to c does consume input, whereas for `a(?=.)c` it does not. So we should indeed ignore characters consumed inside the lookaround, but we shouldn't discard the entire path because of it. Instead we should ignore the contents of the lookaround and continue with the continuation:

```
if (start instanceof LookAroundTree) {
    return canReachWithConsumingInput(start.continuation(), goal, visited);
}
```

(after the `visited.add(start);` line).